### ISSUE TITLE
Understanding with implement php-http and psr-*

### ISSUE THREAD

**Issue by ZhukV** [getsentry/sentry-php#1309]
In library we must provide http client with php-http and request/stream factories with psr-17. I think what this hard to understanding, because php-http full support `psr-7, psr-17, psr-18` and contain specific logic for search http client. But in more cases, projects initialized with one specific http client (Guzzle 7) as an example. And why i MUST install additional php-http package, if i can to try create HttpTransport only with `prs-17, psr-18`?

As result, I see next solution - not require `php-http`, but require `psr/http-client, psr/http-factory` and implement to it. As result, all cleared, developer must create only http client what implement `psr/http-client`. 

If we want to make an easy library to integration, we can move `php-http` to suggestions and prepare builder who was be work via discovery.

Thank for understanding.

**Comment by stayallive** [comment]
I find it a little hard to follow your story but I'm going to do my best to answer but please let me know if I misunderstood anything!

I think you are looking at this package (`sentry/sentry` on Packagist) which needs the PSR implementations to be present in the project.

However we have a so-called meta package which provides all those dependencies which should work out put the box, it's `sentry/sdk` on Packagist which we also recommend installing from our documentation: https://docs.sentry.io/platforms/php/#install.

So the implementation are required (otherwise we cannot send events) and thus they cannot be moved to the suggest, but using the `sentry/sdk` package this should all be provided already.

---

I might have understood your question backwards and you are using `sentry/sdk` but don't want to use the provided libraries there in that case you can remove `sentry/sdk` and require `sentry/sentry` instead and provide your own packages that we need, see: https://docs.sentry.io/platforms/php/configuration/transports/.

**Comment by ZhukV** [comment]
Yes, you are correct, but i want to use `sentry/sentry` packages (not sdk), and why i must installs next packages (required in `composer.json`):

* `php-http/client-common`
* `php-http/discovery`
* `php-http/httpplug`
* `php-http/message`
* `...`

The original `\Sentry\Transport\HttpTransport` can require http client from `psr-18` implementation, but not from `php-http`. But, `sdk` package can contain all additionals for easy integration.

I think previously what this library send async requests (it not specify in psr-18), but package always send synchronous requests.

P.S. As an example. In my projects i full install Guzzle 7. It full support psr-7, psr-17, psr-18. Why i must install `sentry/sdk` or additional packages `php-http`,  if i can to create http transport only with usage Guzzle.

**Comment by stayallive** [comment]
Maybe @Jean85 and/or @ste93cry have a good workaround for you. 

But as far as I understand it we need those implementations for our own transport to work but if you are not using our transport it could make sense that some are not needed, not sure if moving them to the suggest is the correct move though so maybe there are some composer configurations where you can override those requirements.

**Comment by Jean85** [comment]
If you're already installing an HTTP client, you're not required to use `sentry/sdk`, you can use `sentry/sentry` directly.
As for the PHP-HTTP packages, we cannot avoid requiring them, because they are used to autodiscover the transport. With them, you're not required to write any line of code, Guzzle will be detected and used automatically.

If you prefer to wire that manually you still can, as you said, but if we remove the PHP-HTTP requirements we would break the package for everyone else.
